In [2]:
import numpy as np
from os import listdir
import os.path as osp
import pandas as pd

path = "data/register.xls"
labels = pd.read_excel(path)
labels

学習者のID 学習者の性別 学習環境            作文テーマ 学習者の母語   日本語学習履歴 日本語レベル テスト成績（文字語彙）  \
0    CG009      女   国外  外国語がうまくなる方法について    中国語      2年未満     初級           B   
1    CG011      女   国外  外国語がうまくなる方法について    中国語      2年未満     初級           B   
2    CG013      男   国外  外国語がうまくなる方法について    中国語      2年未満     上級           A   
3    CG015      男   国外  外国語がうまくなる方法について    中国語      2年未満     初級           B   
4    CG017      女   国外  外国語がうまくなる方法について    中国語      2年未満     中級           B   
..     ...    ...  ...              ...    ...       ...    ...         ...   
299  KN303      女   国内  外国語がうまくなる方法について    韓国語      2年未満     中級           X   
300  KN307      女   国内  外国語がうまくなる方法について    韓国語  2年以上5年未満     上級           X   
301  KN312      女   国内  外国語がうまくなる方法について    韓国語  2年以上5年未満     上級           X   
302  KN313      男   国内  外国語がうまくなる方法について    韓国語  2年以上5年未満     上級           X   
303  KN316      女   国内  外国語がうまくなる方法について    韓国語      5年以上     上級           X   

    テスト成績（文法）  
0           C  
1           B  
2           A  
3           B  
4           A  
..        ...  
299         X  
300         X  
301         X  
302         X  
303         X  

[304 rows x 9 columns]

In [3]:
from collections import Counter
a = Counter(labels.日本語レベル)
a

Counter({'初級': 31, '上級': 124, '中級': 149})

In [4]:
import numpy as np
label1 = labels.loc[labels.作文テーマ == "外国語がうまくなる方法について",:]
label1

学習者のID 学習者の性別 学習環境            作文テーマ 学習者の母語   日本語学習履歴 日本語レベル テスト成績（文字語彙）  \
0    CG009      女   国外  外国語がうまくなる方法について    中国語      2年未満     初級           B   
1    CG011      女   国外  外国語がうまくなる方法について    中国語      2年未満     初級           B   
2    CG013      男   国外  外国語がうまくなる方法について    中国語      2年未満     上級           A   
3    CG015      男   国外  外国語がうまくなる方法について    中国語      2年未満     初級           B   
4    CG017      女   国外  外国語がうまくなる方法について    中国語      2年未満     中級           B   
..     ...    ...  ...              ...    ...       ...    ...         ...   
299  KN303      女   国内  外国語がうまくなる方法について    韓国語      2年未満     中級           X   
300  KN307      女   国内  外国語がうまくなる方法について    韓国語  2年以上5年未満     上級           X   
301  KN312      女   国内  外国語がうまくなる方法について    韓国語  2年以上5年未満     上級           X   
302  KN313      男   国内  外国語がうまくなる方法について    韓国語  2年以上5年未満     上級           X   
303  KN316      女   国内  外国語がうまくなる方法について    韓国語      5年以上     上級           X   

    テスト成績（文法）  
0           C  
1           B  
2           A  
3           B  
4           A  
..        ...  
299         X  
300         X  
301         X  
302         X  
303         X  

[192 rows x 9 columns]

In [5]:
b = Counter(label1.日本語レベル)
b

Counter({'初級': 31, '上級': 61, '中級': 100})

In [6]:
label2 = labels.loc[labels.作文テーマ == "インターネット時代に新聞や雑誌は必要か",:]
label2

学習者のID 学習者の性別 学習環境                作文テーマ 学習者の母語   日本語学習履歴 日本語レベル  \
76   CG101      男   国外  インターネット時代に新聞や雑誌は必要か    中国語  2年以上5年未満     中級   
77   CG102      女   国外  インターネット時代に新聞や雑誌は必要か    中国語  2年以上5年未満     中級   
78   CG103      女   国外  インターネット時代に新聞や雑誌は必要か    中国語  2年以上5年未満     中級   
79   CG104      女   国外  インターネット時代に新聞や雑誌は必要か    中国語  2年以上5年未満     中級   
80   CG105      女   国外  インターネット時代に新聞や雑誌は必要か    中国語  2年以上5年未満     上級   
..     ...    ...  ...                  ...    ...       ...    ...   
292  KG151      女   国外  インターネット時代に新聞や雑誌は必要か    韓国語      5年以上     上級   
293  KG152      女   国外  インターネット時代に新聞や雑誌は必要か    韓国語  2年以上5年未満     上級   
294  KG153      女   国外  インターネット時代に新聞や雑誌は必要か    韓国語  2年以上5年未満     上級   
295  KG154      女   国外  インターネット時代に新聞や雑誌は必要か    韓国語      2年未満     中級   
296  KG155      女   国外  インターネット時代に新聞や雑誌は必要か    韓国語      5年以上     上級   

    テスト成績（文字語彙） テスト成績（文法）  
76            X         X  
77            X         X  
78            X         X  
79            X         X  
80            X         X  
..          ...       ...  
292           X         X  
293           X         X  
294           X         X  
295           X         X  
296           X         X  

[112 rows x 9 columns]

In [7]:
c = Counter(label2.日本語レベル)
c

Counter({'中級': 49, '上級': 63})

In [8]:
level1 = np.array(label1.日本語レベル)
level1, len(level1)

(array(['初級', '初級', '上級', '初級', '中級', '中級', '中級', '上級', '中級', '中級', '中級',
        '上級', '中級', '中級', '中級', '初級', '中級', '上級', '初級', '初級', '中級', '中級',
        '初級', '中級', '中級', '上級', '中級', '中級', '初級', '上級', '上級', '初級', '中級',
        '中級', '上級', '中級', '中級', '初級', '初級', '中級', '上級', '中級', '上級', '中級',
        '上級', '中級', '初級', '中級', '中級', '初級', '上級', '初級', '中級', '中級', '上級',
        '初級', '初級', '上級', '上級', '中級', '中級', '中級', '上級', '中級', '中級', '上級',
        '中級', '中級', '中級', '初級', '中級', '初級', '中級', '中級', '上級', '上級', '初級',
        '上級', '中級', '中級', '上級', '中級', '中級', '中級', '中級', '中級', '中級', '中級',
        '中級', '中級', '上級', '中級', '中級', '初級', '上級', '上級', '上級', '上級', '上級',
        '中級', '中級', '中級', '中級', '初級', '中級', '中級', '初級', '中級', '中級', '中級',
        '上級', '中級', '上級', '上級', '上級', '上級', '上級', '上級', '上級', '中級', '中級',
        '上級', '中級', '上級', '上級', '中級', '上級', '上級', '中級', '中級', '初級', '中級',
        '中級', '上級', '中級', '上級', '上級', '上級', '中級', '上級', '中級', '上級', '上級',
        '上級', '中級', '初級', '中級', '中級', 

In [9]:
level2 = np.array(label2.日本語レベル)
level2, len(level2)

(array(['中級', '中級', '中級', '中級', '上級', '中級', '中級', '中級', '上級', '中級', '中級',
        '中級', '中級', '上級', '上級', '中級', '中級', '中級', '上級', '中級', '中級', '中級',
        '中級', '上級', '中級', '中級', '中級', '中級', '中級', '上級', '中級', '中級', '中級',
        '上級', '中級', '中級', '中級', '中級', '中級', '中級', '中級', '上級', '中級', '中級',
        '中級', '中級', '上級', '中級', '中級', '上級', '中級', '中級', '中級', '上級', '中級',
        '上級', '上級', '上級', '上級', '中級', '上級', '上級', '上級', '中級', '上級', '上級',
        '中級', '上級', '上級', '上級', '中級', '上級', '上級', '上級', '上級', '上級', '上級',
        '上級', '上級', '上級', '上級', '上級', '上級', '中級', '上級', '上級', '上級', '上級',
        '上級', '上級', '上級', '上級', '上級', '上級', '上級', '上級', '上級', '上級', '上級',
        '上級', '上級', '上級', '上級', '上級', '上級', '上級', '上級', '上級', '上級', '上級',
        '中級', '上級'], dtype=object),
 112)

In [10]:
from os import listdir
txt_path = "data/txt/"
txt_topics = listdir(txt_path)

txt_gaigo_path = txt_path + txt_topics[0]
txt_internet_path = txt_path + txt_topics[1]

txt_gaigo_files = listdir(txt_gaigo_path)
txt_internet_files = listdir(txt_internet_path)

In [11]:
# import os.path as osp
# import re
# gaigo_files = [f for f in listdir(osp.join(txt_path, txt_topics[0])) if f.endswith('.txt')]
# gaigo = []
# for gaigo_file in gaigo_files:
#     # print(gaigo_file)
#     with open(osp.join(txt_path, txt_topics[0], gaigo_file), "r", encoding="utf-8") as f:
#         lines = f.read()
#         # type(lines)
#         lines = re.sub('\n', '', lines)
#         lines = re.sub('\u3000', '', lines)
#     gaigo.append(lines)
# # type(gaigo[0])
# # gaigo[0][0]

import os.path as osp
import re
gaigo_files = [f for f in listdir(osp.join(txt_path, txt_topics[0])) if f.endswith('.txt')]
gaigo = []
for gaigo_file in gaigo_files:
    # print(gaigo_file)
    with open(osp.join(txt_path, txt_topics[0], gaigo_file), "r", encoding="utf-8") as f:
        lines = f.read()
        # type(lines)
        lines = re.sub(r'[0-9０-９a-zA-Zａ-ｚＡ-Ｚ]+', " ", lines)
        lines = re.sub(r'[\．_－―─！＠＃＄％＾＆\-‐|\\＊\“（）＿■×+α※÷⇒—●★☆〇◎◆▼◇△□(：〜～＋=)／*&^%$#@!~`){}・［］…\[\]\"\'\”\’:;<>?＜＞〔〕〈〉？、。・,\./『』【】「」→←○《》≪≫\n\u3000]+', "", lines)
        lines = lines.replace(" ", "")


    gaigo.append(lines)
# type(gaigo[0])
len(gaigo)

# 注意。注釈は削除していない。記号のみ

192

In [12]:
# import re
# internet_files = [f for f in listdir(osp.join(txt_path, txt_topics[1])) if f.endswith('.txt')]
# internet = []
# for internet_file in internet_files:
#     with open(osp.join(txt_path, txt_topics[1], internet_file), "r", encoding="utf-8") as f:
#         lines = f.read()
#         # print(lines)
#         lines = lines.strip('\ufeff□')
#         lines = re.sub('■', '', lines)
#         lines = re.sub('\n', '', lines)
#         lines = re.sub('□', '', lines)
#         lines = re.sub('\u3000', '', lines)
        
#     internet.append(lines)

# internet[0]

import re
internet_files = [f for f in listdir(osp.join(txt_path, txt_topics[1])) if f.endswith('.txt')]
internet = []
for internet_file in internet_files:
    with open(osp.join(txt_path, txt_topics[1], internet_file), "r", encoding="utf-8") as f:
        lines = f.read()
        # print(lines)
        lines = lines.strip('\ufeff□')
        lines = re.sub(r'[0-9０-９a-zA-Zａ-ｚＡ-Ｚ]+', " ", lines)
        lines = re.sub(r'[\．_－―─！＠＃＄％＾＆\-‐|\\＊\“（）＿■×+α※÷⇒—●★☆〇◎◆▼◇△□(：〜～＋=)／*&^%$#@!~`){}・［］…\[\]\"\'\”\’:;<>?＜＞〔〕〈〉？、。・,\./『』【】「」→←○《》≪≫\n\u3000]+', "", lines)
        lines = lines.replace(" ", "")
        
    internet.append(lines)

len(internet)
# + delete ・・・

112

In [13]:
from collections import defaultdict as dd
text_tagged = dd(list)
for level, txt in zip(level1, gaigo):
    text_tagged[level].append(txt)

for level, txt in zip(level2, internet):
    text_tagged[level].append(txt)

text_tagged_dict = dict(text_tagged)
# text_tagged_dict
for level in text_tagged_dict.keys():
    print(len(text_tagged_dict[level]))

31
124
149


In [14]:
for level in text_tagged_dict.keys():
    text_tagged_dict[level] = "".join(text_tagged_dict[level])
        
# text_tagged_dict

In [15]:
import pickle 
# with open('text_tagged_dict2_joined.pickle', 'wb') as f:
#     pickle.dump(text_tagged_dict, f)
    
text_tagged_dict2_joined = pickle.load(open('text_tagged_dict2_joined.pickle', 'rb'))
# text_tagged_dict2_joined

In [21]:
# import matplotlib.pyplot as plt
# import numpy as np

# text_length_dist = [len(text) for level in text_tagged_dict2.keys() for text in text_tagged_dict2[level]]
# # [len(text) for text in text_tagged_dict2["初級"]]
# # text_length_dist

# # np.max(text_length_dist) = 914
# # to determine how many paddings we should add 
# plt.hist(text_length_dist, bins = np.arange(0, 1000, 10))
# plt.show()
# # we can determine the max length of texts as 1000 

In [17]:
text_tagged_dict2_joined["初級"][:10]

'私は大学日本語科の一'

In [19]:
text_tagged_joined2 = {"初級": list(), "上級": list(), "中級": list()}
text_tagged_joined2["初級"].append(text_tagged_dict2_joined["初級"])
text_len_advanced = len(text_tagged_dict2_joined["上級"])
text_len_intermediate = len(text_tagged_dict2_joined["中級"])

for i in range(5):
    text_tagged_joined2["上級"].append(text_tagged_dict2_joined["上級"][int(i * text_len_advanced / 5) : int((i + 1) * text_len_advanced / 5)])

for i in range(6):
    text_tagged_joined2["中級"].append(text_tagged_dict2_joined["中級"][int(i * text_len_intermediate / 6) : int((i + 1) * text_len_intermediate / 6)])

# text_tagged_joined2["中級"][5]
# text_tagged_dict2_joined["中級"]

In [20]:
text_len_advanced

72690

In [21]:
text_len_intermediate

84307

In [22]:
for text in text_tagged_joined2.values():
    print(len(text[0]))

13407
14538
14051


In [23]:
from tqdm import tqdm
import spacy

nlp = spacy.load('ja_ginza')

text_vectorized = {}
for level in tqdm(text_tagged_joined2.keys()):
    text_vectorized[level] = []
    for text in text_tagged_joined2[level]:
        tokenized = nlp(text)
        for token in tokenized:
            text_vectorized[level].append(token.vector)

100%|████████████████████████████████████████████████████████████████████████████████████| 3/3 [00:30<00:00, 10.10s/it]


In [24]:
# for i in text_vectorized.values():
#     print(len(i))

7655
41449
48524


In [1]:
import pickle 
# with open('text_vectorized_joined.pickle', 'wb') as f:
#     pickle.dump(text_vectorized, f)
    
text_vectorized2 = pickle.load(open('text_vectorized_joined.pickle', 'rb'))
# text_vectorized2

In [46]:
import numpy as np

window_size = 15
n_sample = 1000 # less than or equal to (len(text_vectorized2["初級"]) - window_size)


text_transformed = {}
for level, vector in text_vectorized2.items():
    text_transformed[level] = []
    start_index_list = np.random.permutation(list(range(len(text_vectorized2[level]) - window_size)))
    for i in range(n_sample):
        start_index = start_index_list[i]
        # print(start_index)
        chunks = []
        for index in range(start_index, start_index + window_size):
            chunks.append(text_vectorized2[level][index])
        text_transformed[level].append(chunks)
    text_transformed[level] = np.array(text_transformed[level])
    
# print(text_transformed)
# text_transformed["初級"]

In [48]:
for level, value in text_transformed.items():
    print(level, value.shape)

初級 (1000, 15, 300)
上級 (1000, 15, 300)
中級 (1000, 15, 300)


In [2]:
import pickle 
# with open('text_transformed_window_size15_sample_size1000.pickle', 'wb') as f:
#     pickle.dump(text_transformed, f)
    
text_transformed2 = pickle.load(open('text_transformed_window_size15_sample_size1000.pickle', 'rb'))

In [3]:
category_vectors = []
for level, vectors in text_transformed2.items():
    for vector in vectors:
        category_vectors.append((level, vector))

In [4]:
import pandas as pd

key, value = zip(*category_vectors)
data2 = pd.DataFrame({'cat': key, 'value': value})

data2.head()

cat                                              value
0  初級  [[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,...
1  初級  [[0.47779918, -0.290014, 0.35474735, -0.094679...
2  初級  [[-0.11235832, -0.04813227, -0.2808526, 0.0092...
3  初級  [[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0,...
4  初級  [[-0.084042855, -0.03873583, 0.013529049, -0.1...

In [5]:
# for i in data2.value:
#     print(i.shape)

In [6]:
test = data2.sample(frac=0.2, random_state=200)
train = data2.drop(test.index)

test.shape, train.shape

((600, 2), (2400, 2))

In [8]:
from sklearn import preprocessing
import numpy as np

le = preprocessing.LabelEncoder()
ohe = preprocessing.OneHotEncoder()

le.fit(data2.cat)
y_train = le.transform(train.cat).reshape(-1, 1)
ohe.fit(y_train)
y_train = ohe.transform(y_train).todense()

X_train = np.array([x for x in train.value])

X_train.shape, y_train.shape

((2400, 15, 300), (2400, 3))

In [11]:
from keras.models import Sequential
from keras.layers import Dense, LSTM

n_rnn = window_size = 15 # 時系列の数
batch_size = 128
epochs = 20  #epochsは、多いほど、精密に学習するが、重くなるため今回は小さくしている
n_mid = 256  # 中間層のニューロン数
data_dim = 300

model_lstm = Sequential()
model_lstm.add(LSTM(n_mid, input_shape=(n_rnn, data_dim)))
model_lstm.add(Dense(data_dim, activation="relu"))
model_lstm.add(Dense(3, activation="softmax"))
model_lstm.compile(loss='categorical_crossentropy', optimizer="adam")
print(model_lstm.summary())

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 lstm (LSTM)                 (None, 256)               570368    
                                                                 
 dense (Dense)               (None, 300)               77100     
                                                                 
 dense_1 (Dense)             (None, 3)                 903       
                                                                 
Total params: 648,371
Trainable params: 648,371
Non-trainable params: 0
_________________________________________________________________
None


In [12]:
model_lstm.compile(loss='categorical_crossentropy', optimizer="adam", metrics=['accuracy'])
model_lstm.fit(X_train, y_train, epochs=epochs)

Epoch 1/20
75/75 [==============================] - 5s 30ms/step - loss: 1.0410 - accuracy: 0.4146
Epoch 2/20
75/75 [==============================] - 2s 29ms/step - loss: 0.9746 - accuracy: 0.4900
Epoch 3/20
75/75 [==============================] - 2s 29ms/step - loss: 0.9483 - accuracy: 0.5171
Epoch 4/20
75/75 [==============================] - 2s 29ms/step - loss: 0.9178 - accuracy: 0.5379
Epoch 5/20
75/75 [==============================] - 2s 30ms/step - loss: 0.8623 - accuracy: 0.5750
Epoch 6/20
75/75 [==============================] - 2s 30ms/step - loss: 0.8410 - accuracy: 0.5792
Epoch 7/20
75/75 [==============================] - 2s 29ms/step - loss: 0.7817 - accuracy: 0.6300
Epoch 8/20
75/75 [==============================] - 2s 30ms/step - loss: 0.7290 - accuracy: 0.6650
Epoch 9/20
75/75 [==============================] - 2s 30ms/step - loss: 0.6839 - accuracy: 0.6837
Epoch 10/20
75/75 [==============================] - 2s 29ms/step - loss: 0.6105 - accuracy: 0.7267
Epoch 11/

In [13]:
le = preprocessing.LabelEncoder()
ohe = preprocessing.OneHotEncoder()

le.fit(data2.cat)
y_test = le.transform(test.cat).reshape(-1, 1)
ohe.fit(y_test)
y_test = ohe.transform(y_test).todense()

X_test = np.array([x for x in test.value])

X_test.shape, y_test.shape

((600, 15, 300), (600, 3))

In [14]:
from sklearn.metrics import accuracy_score
accuracy_score(np.argmax(model_lstm.predict(X_test), axis=1), np.argmax(y_test, axis=1))

19/19 [==============================] - 1s 16ms/step


0.5066666666666667

The accuracy was 50.7%. This needs fine-tuning. Possible suggestions for fine-tuning are grid

In [15]:
np.argmax(model_lstm.predict(X_test), axis=1)

19/19 [==============================] - 0s 14ms/step


array([0, 2, 1, 0, 2, 2, 2, 0, 2, 1, 2, 1, 1, 1, 0, 2, 2, 1, 0, 1, 2, 0,
       0, 0, 0, 1, 1, 2, 1, 0, 0, 2, 2, 2, 0, 1, 1, 1, 1, 2, 1, 1, 2, 2,
       0, 2, 2, 0, 0, 2, 2, 2, 1, 2, 1, 1, 2, 1, 1, 2, 1, 0, 1, 1, 0, 0,
       2, 1, 0, 0, 1, 1, 2, 2, 1, 1, 2, 2, 1, 1, 0, 0, 1, 2, 0, 2, 2, 2,
       0, 2, 2, 2, 1, 1, 0, 0, 1, 2, 2, 1, 2, 1, 0, 1, 1, 1, 0, 2, 1, 2,
       1, 1, 1, 0, 0, 0, 2, 2, 1, 2, 0, 2, 1, 0, 1, 2, 2, 2, 1, 2, 2, 2,
       0, 2, 1, 0, 1, 1, 1, 2, 1, 1, 1, 1, 2, 1, 2, 2, 1, 0, 2, 0, 1, 0,
       0, 2, 0, 1, 0, 0, 0, 0, 1, 2, 1, 1, 2, 1, 2, 1, 1, 1, 1, 0, 1, 2,
       0, 1, 2, 0, 2, 1, 0, 1, 2, 2, 0, 0, 0, 2, 2, 0, 2, 2, 1, 2, 0, 0,
       0, 2, 1, 0, 1, 0, 1, 0, 2, 1, 0, 0, 1, 2, 1, 0, 2, 2, 1, 1, 0, 0,
       0, 0, 2, 1, 0, 0, 1, 2, 2, 2, 2, 1, 0, 2, 0, 2, 2, 0, 1, 1, 0, 0,
       0, 1, 0, 2, 0, 0, 1, 2, 2, 2, 2, 0, 0, 2, 2, 2, 2, 1, 1, 2, 0, 2,
       1, 2, 1, 2, 2, 0, 1, 0, 0, 0, 0, 0, 0, 2, 2, 0, 1, 2, 2, 1, 0, 0,
       1, 2, 2, 1, 1, 0, 1, 0, 1, 2, 2, 0, 0, 2, 0,

In [16]:
np.argmax(y_test, axis=1).reshape(-1)

matrix([[1, 1, 2, 1, 2, 2, 2, 0, 1, 0, 1, 2, 0, 0, 1, 2, 1, 0, 1, 2, 2,
         0, 1, 1, 0, 1, 1, 2, 0, 1, 0, 2, 1, 1, 2, 0, 1, 0, 2, 2, 0, 2,
         0, 1, 0, 1, 2, 2, 1, 2, 2, 0, 0, 2, 2, 0, 2, 0, 1, 2, 2, 0, 2,
         0, 0, 0, 2, 0, 0, 2, 2, 0, 2, 2, 1, 2, 0, 2, 2, 1, 0, 0, 2, 1,
         0, 2, 2, 2, 0, 2, 2, 2, 0, 0, 0, 0, 1, 1, 1, 0, 2, 1, 1, 1, 1,
         1, 1, 2, 0, 2, 0, 1, 0, 0, 1, 2, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1,
         2, 2, 0, 2, 0, 0, 0, 2, 1, 0, 2, 0, 0, 2, 1, 1, 1, 0, 2, 2, 2,
         1, 1, 1, 2, 1, 0, 0, 1, 2, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 2, 0,
         2, 1, 1, 0, 1, 1, 0, 0, 0, 1, 2, 1, 2, 2, 1, 2, 0, 2, 0, 1, 0,
         2, 0, 0, 2, 2, 1, 0, 0, 1, 0, 2, 1, 0, 2, 1, 1, 0, 1, 0, 2, 2,
         0, 2, 0, 1, 0, 2, 1, 1, 1, 0, 1, 1, 2, 1, 0, 1, 0, 1, 2, 1, 2,
         1, 0, 2, 1, 2, 2, 2, 0, 1, 2, 0, 0, 1, 1, 2, 1, 0, 0, 2, 2, 2,
         2, 0, 1, 0, 2, 2, 1, 1, 0, 2, 1, 1, 1, 2, 2, 2, 2, 1, 2, 1, 2,
         1, 0, 1, 2, 2, 1, 0, 2, 1, 1, 2, 2, 1, 0, 0, 2, 1, 0, 0

In [18]:
import numpy as np
from sklearn import preprocessing
from sklearn.model_selection import GridSearchCV
from keras.models import Sequential
# from keras.layers.core import Dense, Activation
from keras.wrappers.scikit_learn import KerasClassifier
from keras.layers import Dense, LSTM, Dropout
# from keras.optimizers import Adam, Adagrad

In [19]:
window_size = 15
n_sample = 1000 # less than or equal to (len(text_vectorized2["初級"]) - window_size)

n_rnn = window_size  # 時系列の数
batch_size = 128
epochs = 20  #epochsは、多いほど、精密に学習するが、重くなるため今回は小さくしている
n_mid = 256  # 中間層のニューロン数
data_dim = 300

input_dim = (n_rnn, data_dim)
output_dim = 3

In [20]:
def language_model(activation="relu", optimizer="adam", hidden_layer_sizes=(100, 100)):
    model = Sequential()
    firstflag = True
    for dim in hidden_layer_sizes:
        if firstflag:
            model.add(LSTM(n_mid, input_shape=(n_rnn, data_dim), return_sequences=True, activation=activation))
            model.add(Dropout(0.2))
            firstflag = False
        else:
            model.add(LSTM(n_mid, return_sequences=True, activation=activation))
            model.add(Dropout(0.2))

    model.add(LSTM(n_mid, activation=activation))
    model.add(Dropout(0.2))
    model.add(Dense(output_dim, activation="softmax"))
    model.compile(loss='categorical_crossentropy', optimizer=optimizer, metrics=['accuracy'])
    return model

In [21]:
activation = ["relu", "tanh"]
optimizer = ["adam", "adagrad", "sgd", "RMSprop", "Adamax"]
hidden_layer_sizes = [(50, 50, 50, 50), (50, 50, 50), (50, 50), (50, ) ]
nb_epoch = [20, 25]
batch_size = [128, 256]

In [22]:
model2 = KerasClassifier(build_fn=language_model, verbose=0)
param_grid = dict(activation=activation, 
                  optimizer=optimizer, 
                  hidden_layer_sizes=hidden_layer_sizes, 
                  nb_epoch=nb_epoch, 
                  batch_size=batch_size,)
grid = GridSearchCV(estimator=model2, param_grid=param_grid, verbose=2)


C:\Users\fic2023150.FIC.009\AppData\Local\Temp\ipykernel_2724\2440950011.py:1: DeprecationWarning: KerasClassifier is deprecated, use Sci-Keras (https://github.com/adriangb/scikeras) instead. See https://www.adriangb.com/scikeras/stable/migration.html for help migrating.
  model2 = KerasClassifier(build_fn=language_model, verbose=0)


In [23]:
grid_result = grid.fit(X_train, y_train)

Fitting 5 folds for each of 160 candidates, totalling 800 fits
[CV] END activation=relu, batch_size=128, hidden_layer_sizes=(50, 50, 50, 50), nb_epoch=20, optimizer=adam; total time=   8.5s
[CV] END activation=relu, batch_size=128, hidden_layer_sizes=(50, 50, 50, 50), nb_epoch=20, optimizer=adam; total time=  13.4s
[CV] END activation=relu, batch_size=128, hidden_layer_sizes=(50, 50, 50, 50), nb_epoch=20, optimizer=adam; total time=  16.3s
[CV] END activation=relu, batch_size=128, hidden_layer_sizes=(50, 50, 50, 50), nb_epoch=20, optimizer=adam; total time=  18.4s
[CV] END activation=relu, batch_size=128, hidden_layer_sizes=(50, 50, 50, 50), nb_epoch=20, optimizer=adam; total time=  22.7s
[CV] END activation=relu, batch_size=128, hidden_layer_sizes=(50, 50, 50, 50), nb_epoch=20, optimizer=adagrad; total time=  23.4s
[CV] END activation=relu, batch_size=128, hidden_layer_sizes=(50, 50, 50, 50), nb_epoch=20, optimizer=adagrad; total time=  25.2s
[CV] END activation=relu, batch_size=128, 

KeyboardInterrupt: 

In [68]:
grid_result

GridSearchCV(estimator=<keras.wrappers.scikit_learn.KerasClassifier object at 0x000001B692177970>,
             param_grid={'activation': ['relu', 'sigmoid'],
                         'batch_size': [5, 10],
                         'hidden_layer_sizes': [(50, 50, 50, 50), (50, 50, 50),
                                                (100, 100), (50,)],
                         'nb_epoch': [10, 25],
                         'optimizer': ['adam', 'adagrad', 'sgd', 'RMSprop',
                                       'Adamax']},
             verbose=2)

In [69]:
grid_result.best_params_

{'activation': 'relu',
 'batch_size': 5,
 'hidden_layer_sizes': (50,),
 'nb_epoch': 25,
 'optimizer': 'RMSprop'}

In [24]:
activation = ["tanh"]
optimizer = ["adam", "RMSprop"]
hidden_layer_sizes = [(50, 50, 50, 50), (50, 50, 50), (50, 50), (50, ) ]
nb_epoch = [20]
batch_size = [128]

In [25]:
model3 = KerasClassifier(build_fn=language_model, verbose=0)
param_grid = dict(activation=activation, 
                  optimizer=optimizer, 
                  hidden_layer_sizes=hidden_layer_sizes, 
                  nb_epoch=nb_epoch, 
                  batch_size=batch_size,)
grid2 = GridSearchCV(estimator=model3, param_grid=param_grid, verbose=2)


C:\Users\fic2023150.FIC.009\AppData\Local\Temp\ipykernel_2724\4269649812.py:1: DeprecationWarning: KerasClassifier is deprecated, use Sci-Keras (https://github.com/adriangb/scikeras) instead. See https://www.adriangb.com/scikeras/stable/migration.html for help migrating.
  model3 = KerasClassifier(build_fn=language_model, verbose=0)


In [26]:
grid_result2 = grid2.fit(X_train, y_train)

Fitting 5 folds for each of 8 candidates, totalling 40 fits
[CV] END activation=tanh, batch_size=128, hidden_layer_sizes=(50, 50, 50, 50), nb_epoch=20, optimizer=adam; total time=  29.8s
[CV] END activation=tanh, batch_size=128, hidden_layer_sizes=(50, 50, 50, 50), nb_epoch=20, optimizer=adam; total time=  33.0s
[CV] END activation=tanh, batch_size=128, hidden_layer_sizes=(50, 50, 50, 50), nb_epoch=20, optimizer=adam; total time=  32.4s
[CV] END activation=tanh, batch_size=128, hidden_layer_sizes=(50, 50, 50, 50), nb_epoch=20, optimizer=adam; total time=  34.1s
[CV] END activation=tanh, batch_size=128, hidden_layer_sizes=(50, 50, 50, 50), nb_epoch=20, optimizer=adam; total time=  32.6s
[CV] END activation=tanh, batch_size=128, hidden_layer_sizes=(50, 50, 50, 50), nb_epoch=20, optimizer=RMSprop; total time=  33.9s
[CV] END activation=tanh, batch_size=128, hidden_layer_sizes=(50, 50, 50, 50), nb_epoch=20, optimizer=RMSprop; total time=  35.5s
[CV] END activation=tanh, batch_size=128, hid

In [28]:
grid_result2

GridSearchCV(estimator=<keras.wrappers.scikit_learn.KerasClassifier object at 0x00000204DE286F70>,
             param_grid={'activation': ['tanh'], 'batch_size': [128],
                         'hidden_layer_sizes': [(50, 50, 50, 50), (50, 50, 50),
                                                (50, 50), (50,)],
                         'nb_epoch': [20], 'optimizer': ['adam', 'RMSprop']},
             verbose=2)

In [29]:
grid_result2.best_params_

{'activation': 'tanh',
 'batch_size': 128,
 'hidden_layer_sizes': (50, 50),
 'nb_epoch': 20,
 'optimizer': 'adam'}

In [30]:
grid_result2.best_params_['activation']

'tanh'

In [32]:
model4 = language_model(activation=grid_result2.best_params_['activation'], 
                   optimizer=grid_result2.best_params_['optimizer'], 
                   hidden_layer_sizes=grid_result2.best_params_['hidden_layer_sizes'])

model4.fit(X_train, y_train, epochs=grid_result2.best_params_['nb_epoch'], batch_size=grid_result2.best_params_['batch_size'])

Epoch 1/20
19/19 [==============================] - 25s 1s/step - loss: 1.0656 - accuracy: 0.4283
Epoch 2/20
19/19 [==============================] - 22s 1s/step - loss: 1.0143 - accuracy: 0.4812
Epoch 3/20
19/19 [==============================] - 22s 1s/step - loss: 0.9789 - accuracy: 0.4883
Epoch 4/20
19/19 [==============================] - 22s 1s/step - loss: 0.9697 - accuracy: 0.4888
Epoch 5/20
19/19 [==============================] - 22s 1s/step - loss: 0.9405 - accuracy: 0.5225
Epoch 6/20
19/19 [==============================] - 22s 1s/step - loss: 0.9164 - accuracy: 0.5379
Epoch 7/20
19/19 [==============================] - 22s 1s/step - loss: 0.8936 - accuracy: 0.5412
Epoch 8/20
19/19 [==============================] - 22s 1s/step - loss: 0.8651 - accuracy: 0.5750
Epoch 9/20
19/19 [==============================] - 22s 1s/step - loss: 0.8640 - accuracy: 0.5633
Epoch 10/20
19/19 [==============================] - 22s 1s/step - loss: 0.8880 - accuracy: 0.5554
Epoch 11/20
19/19 [

In [33]:
from sklearn.metrics import accuracy_score
# accuracy_score(model2.predict(test_X), test_y)
accuracy_score(np.argmax(model4.predict(X_test), axis=1), np.argmax(y_test, axis=1))

19/19 [==============================] - 3s 106ms/step


0.53

In [37]:
np.argmax(model4.predict(X_test), axis=1), np.argmax(y_test, axis=1).reshape(-1)

19/19 [==============================] - 2s 104ms/step


(array([0, 2, 2, 2, 2, 0, 2, 0, 2, 2, 2, 2, 2, 0, 0, 2, 2, 2, 2, 1, 2, 1,
        0, 0, 0, 2, 0, 2, 0, 1, 0, 2, 2, 2, 2, 0, 2, 0, 2, 2, 2, 2, 2, 2,
        0, 2, 2, 0, 1, 2, 2, 2, 0, 2, 1, 0, 2, 0, 0, 2, 1, 0, 2, 0, 0, 0,
        2, 0, 0, 2, 2, 1, 2, 2, 0, 2, 2, 2, 2, 1, 0, 0, 1, 0, 2, 2, 2, 2,
        0, 2, 2, 2, 1, 2, 0, 0, 1, 2, 2, 0, 2, 0, 2, 0, 2, 1, 0, 2, 1, 2,
        0, 0, 0, 0, 0, 2, 2, 2, 0, 2, 0, 2, 0, 0, 2, 2, 2, 2, 1, 2, 2, 2,
        1, 2, 1, 0, 1, 1, 2, 2, 2, 0, 0, 1, 1, 2, 2, 2, 1, 0, 0, 0, 0, 0,
        0, 0, 0, 1, 0, 1, 0, 0, 1, 2, 1, 2, 2, 1, 2, 0, 0, 0, 1, 0, 1, 1,
        0, 2, 2, 0, 2, 1, 0, 2, 2, 2, 1, 0, 0, 2, 2, 0, 2, 2, 2, 2, 0, 0,
        2, 2, 0, 0, 2, 1, 0, 0, 2, 0, 0, 0, 1, 2, 0, 0, 2, 2, 1, 1, 0, 0,
        2, 0, 2, 0, 0, 0, 1, 2, 2, 1, 2, 2, 0, 2, 0, 2, 2, 2, 1, 1, 2, 0,
        0, 2, 1, 2, 0, 0, 0, 2, 2, 0, 2, 0, 0, 1, 2, 2, 1, 1, 1, 2, 2, 2,
        0, 2, 2, 2, 2, 1, 2, 1, 0, 0, 0, 0, 2, 2, 2, 0, 1, 1, 2, 2, 0, 1,
        0, 2, 2, 1, 2, 0, 0, 0, 0, 2, 

In [ ]:
# for early stopping
# https://qiita.com/sasayabaku/items/b7872a3b8acc7d6261bf

# stacked lstm model example
# https://machinelearningknowledge.ai/keras-lstm-layer-explained-for-beginners-with-example/

# lstm layer documentation
# https://keras.io/api/layers/recurrent_layers/lstm/

# Should you use relu for activation in lstm
# https://stats.stackexchange.com/questions/444923/activation-function-between-lstm-layers

# lstm explained
# https://qiita.com/t_Signull/items/21b82be280b46f467d1b

# stacked lstm explained
# https://machinelearningmastery.com/stacked-long-short-term-memory-networks/

# error solved
# https://stackoverflow.com/questions/58119320/valueerror-input-0-of-layer-lstm-is-incompatible-with-the-layer-expected-ndim

In [38]:
model5 = language_model(activation="tanh", 
                   optimizer='adagrad', 
                   hidden_layer_sizes=(50,))

model5.fit(X_train, y_train, epochs=20, batch_size=256)

Epoch 1/20
10/10 [==============================] - 14s 1s/step - loss: 1.1014 - accuracy: 0.3075
Epoch 2/20
10/10 [==============================] - 12s 1s/step - loss: 1.1005 - accuracy: 0.3171
Epoch 3/20
10/10 [==============================] - 12s 1s/step - loss: 1.1006 - accuracy: 0.3083
Epoch 4/20
10/10 [==============================] - 12s 1s/step - loss: 1.1009 - accuracy: 0.3021
Epoch 5/20
10/10 [==============================] - 12s 1s/step - loss: 1.1005 - accuracy: 0.3000
Epoch 6/20
10/10 [==============================] - 12s 1s/step - loss: 1.1005 - accuracy: 0.3038
Epoch 7/20
10/10 [==============================] - 12s 1s/step - loss: 1.1002 - accuracy: 0.3179
Epoch 8/20
10/10 [==============================] - 12s 1s/step - loss: 1.1003 - accuracy: 0.3192
Epoch 9/20
10/10 [==============================] - 12s 1s/step - loss: 1.1002 - accuracy: 0.3167
Epoch 10/20
10/10 [==============================] - 12s 1s/step - loss: 1.1002 - accuracy: 0.3021
Epoch 11/20
10/10 [

In [39]:
def language_model2(optimizer="adam", hidden_layer_sizes=(100, 100)):
    model = Sequential()
    firstflag = True
    for dim in hidden_layer_sizes:
        if firstflag:
            model.add(LSTM(n_mid, input_shape=(n_rnn, data_dim), return_sequences=True))
            model.add(Dropout(0.2))
            firstflag = False
        else:
            model.add(LSTM(n_mid, return_sequences=True))
            model.add(Dropout(0.2))

    model.add(LSTM(n_mid))
    model.add(Dropout(0.2))
    model.add(Dense(output_dim, activation="softmax"))
    model.compile(loss='categorical_crossentropy', optimizer=optimizer, metrics=['accuracy'])
    return model

In [40]:
# activation = ["relu", "tanh"]
optimizer = ["adam", "adagrad", "sgd", "RMSprop", "Adamax"]
hidden_layer_sizes = [(50, 50, 50, 50), (50, 50, 50), (50, 50), (50, ) ]
nb_epoch = [20, 25]
batch_size = [128, 256]

In [41]:
model5 = KerasClassifier(build_fn=language_model2, verbose=0)
param_grid = dict(##activation=activation, 
                  optimizer=optimizer, 
                  hidden_layer_sizes=hidden_layer_sizes, 
                  nb_epoch=nb_epoch, 
                  batch_size=batch_size,)
grid3 = GridSearchCV(estimator=model5, param_grid=param_grid, verbose=2)


C:\Users\fic2023150.FIC.009\AppData\Local\Temp\ipykernel_2724\495326255.py:1: DeprecationWarning: KerasClassifier is deprecated, use Sci-Keras (https://github.com/adriangb/scikeras) instead. See https://www.adriangb.com/scikeras/stable/migration.html for help migrating.
  model5 = KerasClassifier(build_fn=language_model2, verbose=0)


In [ ]:
grid3_result = grid3.fit(X_train, y_train)

In [ ]:
## grid_result.cv_results_['mean_test_score']
## https://stackoverflow.com/questions/68844792/lstm-will-not-use-cudnn-kernels-since-it-doesnt-meet-the-criteria-it-will-use
# Colab.research.google.com/drive/1VfGwo6Y8nkCMdiyyze6rPoJ0cmjE7ld3#scrollTo=cf9d0590-7607-4165-8ab9-7254250d7d16